src/firefly/java/edu/caltech/ipac/firefly/server/servlets/AnyFileDownload.java

```diff
@@ -32,71 +31,125 @@
+    private static final String STATIC_CACHE_CONTROL= "max-age=86400";
 
     protected void processRequest(HttpServletRequest req, HttpServletResponse res) throws Exception {
-        SrvParam sp= new SrvParam(req.getParameterMap());
-        String hips= sp.getOptional(HIPS_PARAM);
-        String fname= hips==null ? sp.getRequired(FILE_PARAM): null;
-        String local= sp.getOptional(RETURN_PARAM);
-        boolean log= sp.getOptionalBoolean(LOG_PARAM,false);
-        boolean track= sp.getOptionalBoolean(TRACK_PARAM,false);
-        File downloadFile;
+        if (req.getParameterMap().containsKey(HIPS_PARAM)) {
+            handleHiPSRequest(req,res,true);
+        }
+        else {
+            handleInternalFileRequest(req,res,true);
+        }
+
+    }
 
-        if (hips!=null) {
-            downloadFile= retrieveHiPSData(hips, null);
+    @Override
+    protected void doHead(HttpServletRequest req, HttpServletResponse res) throws IOException {
+        if (req.getParameterMap().containsKey(HIPS_PARAM)) {
+            handleHiPSRequest(req,res,false);
         }
         else {
-            downloadFile= ServerContext.convertToFile(fname);
+            handleInternalFileRequest(req,res,false);
         }
+    }
 
+    private void handleInternalFileRequest(HttpServletRequest req, HttpServletResponse res, boolean fullBody)
+            throws IOException {
+        SrvParam sp= new SrvParam(req.getParameterMap());
+        String fname= sp.getRequired(FILE_PARAM);
+        File downloadFile= ServerContext.convertToFile(fname);
         if (downloadFile==null) {
             res.sendError(HttpServletResponse.SC_NOT_FOUND,"Could not convert input to a file" );
             _log.warn("Cannot convert file: "+ fname + " to valid path, returning 404");
         }
         else if (!downloadFile.canRead()) {
             res.sendError(HttpServletResponse.SC_NOT_FOUND,"File was not found" );
             _log.warn("Cannot read file: "+ downloadFile.getPath(),
-                      "fname: "+fname, " returning 404");
+                    "fname: "+fname, " returning 404");
         }
         else {
-            res.addHeader("Cache-Control", "max-age=86400");
-            String retFile= (local!=null) ? local : downloadFile.getName();
-            if (retFile.equals(USE_SERVER_NAME)) retFile= downloadFile.getName();
-            insertResponseHeaders(res, downloadFile,retFile);
-            if (track) trackProgress(req, BackgroundEnv.DownloadProgress.WORKING);
-            try {
-                FileUtil.writeFileToStream(downloadFile, res.getOutputStream());
-                if (track) trackProgress(req, BackgroundEnv.DownloadProgress.DONE);
-                if (log) logActivity(downloadFile);
-            } catch (IOException e) {
-                if (track) trackProgress(req, BackgroundEnv.DownloadProgress.FAIL);
-                throw e;
+            res.addHeader("Cache-Control", STATIC_CACHE_CONTROL);
+            if (fullBody) {
```

Comment:
Now, I understand the problem when you say, the data cannot be 'refetch' because it was created upstream.
But, I think it should not matter.  At this point, you're simply saying I have the file and it has not changed, or the file has been deleted, don't use your cached version.  

In the case when the file no longer on the server, the UI may show visible 'exceptions' which forces them to recreate the missing files.  In my opinion, it's still better than serving cached data that we cannot control on the server's end.

Reply:
You are saying just remove cache control in this case. This is used for unique file that we have created internally.

Reply:
I did look at how it is used.  Mostly to download packaged files, region file we create and fits files.

Reply:
I am thinking of the png files for webplot viewer.  If it's not going through this code path, then ignore what I said.  For packaged files, like you said, caching does not matter one way or the other.

Reply:
png files are all generated locally now, as of a year ago.